### DISCUSSION TITLE
Solving the wrong calculation of SOC on JK-BMS board

### DISCUSSION THREAD

**Discussion by paulsteigel** [Sleeper85/esphome-jk-bms-can#19] [Ideas]
Hi Sleeper85,
Thank you for your work that I found brilliant and helpful for DIY community. As you knew, Jk-bms SOC calculation has problem over time ([Google Link](https://www.google.com/search?q=jkbms+%2B+soc+problem&rlz=1C1CHBF_enVN981VN981&oq=jkbms+%2B+soc+problem&gs_lcrp=EgZjaHJvbWUyBggAEEUYOTIJCAEQIRgKGKABMgkIAhAhGAoYoAHSAQg5MDI2ajBqN6gCALACAA&sourceid=chrome&ie=UTF-8)).
And it seems that Inverter prefer working with SOC (mine Deye and Luxpower like SOC rather than using voltage). Since this was problem, I make a hybrid getMappedSOC function that based on voltage for LFP battery (can also be applied with lithium ion battery) with power counter (in Watt per Second interval - W2S). This is made to avoid the case that getMappedSOC will fluctuate when current is charging quickly. SOC base on voltage is not good but it can be a basic point for avoiding the wrong SOC calculation of JK-BMS.
The process is described as following:
INPUT: BatteryAh (in Ah)
REQUIRE: Calculate starting remained energy W2S
LOOP: Every 2 seconds 
+ Startup: 
++ Calculate SocAtStart base on Voltage
++ W2S = SocAtStart * 60 * 30
+ Overtime (Looping interval is 2 second)
W2S = W2S + (Voltage * Current): Current can be negative or positive so it cause PowerCounter be increased or decreased.
+ Check if current is:
++ 0A: SocAtIdle = GetMapSOC >> W2S = W2S - (SocAtStart - SocAtIdle) * 60 * 30: This is to help finetuing the SOC as closer to the OCV state as better.
NewSOC = W2S/BatteryAh/60/30.

My GetSocMap function is below:
```
#define LFP_LEN 15
#define LION_LEN 21

uint8_t MapSOCLFP[LFP_LEN] = {0, 1, 5, 10, 14, 20, 30, 40, 50, 60, 70, 80, 90, 99, 100}; // array to keep SOC level
uint8_t MapSOCLion[LION_LEN] = {0, 5, 10, 15, 20, 25, 30, 35, 40, 45, 50, 55, 60, 65, 70, 75, 80, 85, 90, 95, 100};
//uint16_t MapVoltLFP[LFP_LEN] = {2500, 2538, 2800, 3000, 3150, 3200, 3225, 3250, 3263, 3275, 3300, 3325, 3350, 3375, 3450}; // Keep Cell milivolt for LFP mapping
uint16_t MapVoltLFP[LFP_LEN] = {40000, 40608, 44800, 48000, 50400, 51200, 51600, 52000, 52208, 52400, 52800, 53200, 53600, 54000, 55200};
//uint16_t MapVoltLion[LION_LEN] = {3000, 3062, 3123, 3177, 3238, 3300, 3362, 3423, 3477, 3538, 3600, 3662, 3723, 3777, 3838, 3900, 3962, 4023, 4077, 4138, 4200};
uint16_t MapVoltLion[LION_LEN] = {39000, 39800, 40600, 41300, 42100, 42900, 43700, 44500, 45200, 46000, 46800, 47600, 48400, 49100, 49900, 50700, 51500, 52300, 53000, 53800, 54600};

uint8_t getMappedSOC(uint16_t AverageRefVoltage){
  uint16_t RefVoltage = 0;
  uint16_t UpperBound;
  uint16_t LowerBound;
  uint16_t *pMapVol;
  uint8_t *pMapSOC;
  uint8_t fakeSoc;
  uint8_t arrLength;    
  uint8_t NumDivider;
  
  if (JKComputedData.BatteryType == 0) {
    arrLength=LFP_LEN;
    pMapVol = MapVoltLFP;
    pMapSOC = MapSOCLFP;
    NumDivider = 16L;
  }else if (JKComputedData.BatteryType == 1) {
    arrLength=LION_LEN;
    pMapVol = MapVoltLion;
    pMapSOC = MapSOCLion;
    NumDivider = 13L;    
  }
  //LastVoltageMillivolt;//
  RefVoltage = AverageRefVoltage;
  
  for (uint8_t ptr = 3; ptr < arrLength; ptr++){ //Stop at 80%    
    if(RefVoltage >= *(pMapVol + ptr) && RefVoltage <= *(pMapVol + ptr +1)){
      uint16_t UpperBound = *(pMapVol + ptr + 1) * uint32_t(JKComputedData.ActualNumberOfCellInfoEntries) / NumDivider;      
      uint16_t LowerBound = *(pMapVol + ptr) * uint32_t(JKComputedData.ActualNumberOfCellInfoEntries) / NumDivider;
      //Serial.print(F("After calculation = "));Serial.print(LowerBound);Serial.print(F("/"));Serial.println(UpperBound);
      fakeSoc = map(RefVoltage, LowerBound, UpperBound, *(pMapSOC +ptr), *(pMapSOC + ptr + 1));      
      //fakeSoc = fakeSoc + 3;            
      return fakeSoc;
    }
  }
  //Serial.println(F("No map returned!!"));
  return 20;
}
```
And my Arduino code for W2S is below:
```
uint8_t GetSOCPercent(){
  // Get inital SOC at starting SOCAtStarting
  // wherenever battery return to idle state, OCV shall be calculated for updating SOC again
  
  if (W2S == 1111111111) {
    //Serial.print(F("Voltage Start = "));Serial.print(JKComputedData.BatteryVoltage10Millivolt * 10);Serial.println(F("%"));
    SOCAtStarting = getMappedSOC(JKComputedData.BatteryVoltage10Millivolt * 10);
    // Arduino is just started or restarted, get mapped Soc for starting
    //Serial.print(F("Start SOC 1 = "));Serial.print(SOCAtStarting);Serial.println(F("%"));
    W2S = uint32_t(SOCAtStarting) *  JKComputedData.TotalCapacityAmpereHour * 60L * 30L ; // = Ah * SOCAtStarting / 100
    //Serial.print(F("Start W2S = "));Serial.print(W2S / JKComputedData.TotalCapacityAmpereHour / 60L / 30L);
  } else {
    if (JKComputedData.Battery10MilliAmpere == 0 && !IsCalculated){
      IsCalculated = true;
      SOCAtIdle = getMappedSOC(JKComputedData.BatteryVoltage10Millivolt * 10);
      Serial.print(F("SOCAtIdle = "));Serial.println(SOCAtIdle);
      // recalculate W2S
      W2S = W2S - (SOCAtStarting - SOCAtIdle) * 60L * 30L;
    }          
    //Serial.print(F("var1 * var2 = "));Serial.println(JKComputedData.BatteryVoltage10Millivolt * 10 * JKComputedData.Battery10MilliAmpere * 10 / 1000L / 1000L);
    W2S += (int32_t(JKComputedData.BatteryVoltage10Millivolt) * int32_t(JKComputedData.Battery10MilliAmpere) / 100L / 100L);
  }
  
  //Serial.print(F("WattPer2Seconds = "));Serial.print(W2S);Serial.print(F("New SOC = "));Serial.print(W2S / JKComputedData.TotalCapacityAmpereHour / 60L / 30L);Serial.println(F("%"));
  // Process to control on over/undervoltage/unbalanced and absorption charging
  if ((millis() - LastCheckTime) > CHARGE_STATUS_REFRESH_INTERVAL) {
    LastCheckTime = millis();
    //Stage 1: Check for DeltaVolt limit
    if (JKComputedData.DeltaCellMillivolt >= DELTA_VOLT_RANGE) {
      // Start to reduce charging current
      MinuteCount ++;
      ReduceCurrent(1);          
    } else {
      resetCharge();
      // Set new current limit to the previous safe limit at stage 1
      if (Computed_Current_limits_100mA == 0) Computed_Current_limits_100mA = JKComputedData.Battery10MilliAmpere / 10L;   
      if (Charge_Current_100_milliAmp != 0) Computed_Current_limits_100mA = Charge_Current_100_milliAmp; // set new current value limit 
      //Serial.print(F("Computed_Current_limits_100mA:"));Serial.println(Computed_Current_limits_100mA);
      //Serial.print(F("Charge_Current_100_milliAmp:"));Serial.println(Computed_Current_limits_100mA);
      // reset counter and continue with set current      
      MinuteCount = 1;      
    }
    // Stage 2: Check for Over voltage protection        
    if (JKComputedData.MaximumCellMillivolt * 1.03 >= JKComputedData.CellOvervoltageProtectionMillivolt){
      Serial.print(F("OVP detected!!"));
      MinuteCount ++;
      ReduceCurrent(2);
    }else {
      resetCharge();
      MinuteCount = 0;
    }
    // Stage 3: Check for under voltage protection
    if (JKComputedData.MinimumCellMillivolt * 0.95 <= JKComputedData.CellUndervoltageProtectionMillivolt){
      Serial.print(F("UVP detected!!"));
      // Stop discharging
    }else {
      resetCharge();
      MinuteCount = 0;
    }
  }
  //float tBatteryToFullDifference = JKComputedData.BatteryFullVoltage10Millivolt - JKComputedData.BatteryVoltage10Millivolt;  
  //Serial.print(F("SOCPercent = "));Serial.println(W2S / JKComputedData.TotalCapacityAmpereHour / 60L / 30L);
  return W2S / JKComputedData.TotalCapacityAmpereHour / 60L / 30L;
}
```
With this approach, I believe, the ESP32 will act as both the CAN communicator and SOC calculator (base on power and voltage) for the inverter to work better.
Can you advise on this perspective with thanks.
Ngoc.

**Comment by Sleeper85** [comment]
Hi,

Indeed, JK-BMS is not very good at calculating the SoC especially during winter when there is not 100% recharge every day.

I don't have time to analyze your function now. I'm about to leave for a 4 month trip and I won't be able to do any more tests.

But I'm open to discussing it with you.

All I can say is that to calculate a good SoC you need a good source of information like a shunt. I don't see how to calculate it better than JK-BMS if we base ourselves on the voltage, current, power information from JK-BMS.

The problem of JK-BMS is its miscalculation of the current when it is < 2A. This causes a lot of calculation errors during the night when there is little consumption. I calibrate it for a good measurement of low current.

There is some code regarding the SoC in the latest version because many inverters stop charging when the SoC reaches 100%.

```c++
// SOC - Sending 100% only at the end of the absorption phase
if (id(state_of_charge).state < 100) soc = id(state_of_charge).state;       // SOC < 100% => Sending BMS SOC
else if (id(eoc) == true) soc = 100;                                        // End Of Charge => Sending 100%
else soc = 99;                                                              // Otherwise => Sending 99%
```

I will soon release the new YAML version 1.17.2 with lots of new things.

- V1.17.2 MrPablo   : Added function "Auto Charge/Discharge Current Control" to avoid OVP/UVP alarms
- V1.17.1 Sleeper85 : New Cut-off Current/Voltage Charging Logic for LFP with the participation of Shvm
- V1.16.6 Sleeper85 : Selectable CAN settings + Adding inverter_offset_v + Improved CAN ID 0x355, sending 100% only at the end of the absorption phase, adding bytes [04:05] and [06:07] + Automatic calculation of the number of battery modules + Save and Restore slider values
- V1.16.5 Sleeper85 : Add Preventive Alarms Logic, CAN ID 0x356: send average temperature of T1/T2, new "Discharging current max"

Have a nice day ;-)

**Comment by paulsteigel** [comment]
Hi,
Great thanks to your reply.
I do agree with you on JK SOC calculation. That was why I thought of a fake
SOC solution. Many Vietnamese are using JK as they are cheap and
affordable. In fact, I tested the fakeSOC version only (only provide SOC
base on voltage) and it successfully tells the Inverter to work properly
based on voltage that was reflected by fake SOC. The only drawback of this
solution is the fluctuation of SOC at charge/discharge at high current.
The new version (I shared over the discussion) was tested in few days and
need more refinement. That why I would like to raise an idea with you!
Many thanks for your response and wish you a nice and beautiful trip.
Ngoc

**Comment by straga** [comment]
I found some information from vicrton dbus.

https://github.com/Louisvdw/dbus-serialbattery/blob/master/etc/dbus-serialbattery/config.default.ini

This will add a feature to automatically reset SOC of JK BMS based on the manage_charge_voltage calculations.
**Warning:** This will modify OVP and OVPR settings of your BMS when enabled. **Use it at your own risk!**

AUTO_RESET_SOC and JK_BMS_AUTO_RESET_SOC have to be enabled in your config to make it work.

Code for that
`   
 async def reset_soc_jk(self, c):

        # Lowering OVPR / OVP based on the maximum cell voltage at the time
        # That will trigger a High Voltage Alert and resets SOC to 100%
        ovp_trigger = round(self.max_cell_voltage - 0.05, 3)
        ovpr_trigger = round(self.max_cell_voltage - 0.10, 3)
        await self.write_register(
            JK_REGISTER_OVPR, self.jk_float_to_hex_little(ovpr_trigger), 0x04, c, True
        )
        await self.write_register(
            JK_REGISTER_OVP, self.jk_float_to_hex_little(ovp_trigger), 0x04, c, True
        )
        # Give BMS some time to recognize
        await asyncio.sleep(5)

        # Set values back to initial values
        await self.write_register(
            JK_REGISTER_OVP,
            self.jk_float_to_hex_little(self.ovp_initial_voltage),
            0x04,
            c,
            True,
        )
        await self.write_register(
            JK_REGISTER_OVPR,
            self.jk_float_to_hex_little(self.ovpr_initial_voltage),
            0x04,
            c,
            True,
        )

        logging.info("JK BMS SOC reset finished.")
`

**Comment by paulsteigel** [comment]
Hi Victor,
Many thanks for your sharing. Actually, I has passed by the discussion
before thinking of working on the getFakeSoc solution. This is because, the
config required bluetooth connection not normal UART/Serial connection
(impossible for making change to the JK-BMS via serial).
I will ask Mr. Manuel on how to deal with the matter!
Best
Ngoc

**Comment by Sleeper85** [comment]
But the best would still be to use a shunt as a source. Several of us had this idea, including myself. I have a Junctek Shunt (good value for money) that I have already integrated into HA and the 2 other interested people (@MrPabloUK) have a Victron Shunt.

**Comment by straga** [comment]
Yes,  Junctek Shunt the best solution from my point.

**Comment by paulsteigel** [comment]
Yeah, you are all totally correct. However, as I have been observing the community in our places, most people just use JKBMS with battery and connect to the inverter as a lead acid battery (that use voltage as indicator for deciding charge/discharge). For the moment, this choice is mandated and prevail for all people here with DIY battery.
 JK-BMS is cheap and affordable, have some way to let the battery talk to the inverter is awesome but to invest some 100$ to buy a Smart Shunt (Victron/ Junctek ...) seems not the choice of many in our place.
This is why I am still sleeping on the option of how to smoothly converting the voltage to some concept that inverter accept for treating the battery nicely with maximum durability with least investment (20-30$ or below on an esp32 and few other parts).
Best
Ngoc

**Comment by kommando828** [comment]
The problem with using Voltage alone to determine SOC with LFP batteries is the very flat section in the Voltage/SOC curve where the Voltage hardly varies for a large range of SOC. Hence why the Shunt is a good solution as they record the AH in and out which fills in the gap at the flat section of the curve but does come with an expense especially the Victron,. That does not mean there is not a Shuntless solution, just that it needs some other input than Voltage alone and also does not need 100% SOC to be achieved daily. Its a pity low cost AC CT clamps can't be used but DC versions cost more than a shunt.

**Reply by paulsteigel** [reply]
Absolutely Kommando828,
As user here thay just want to make sure (here is my case when i configured LFP battery as lead acid in the Inverter):
1. Cells are not unbalanced too much;
2. Voltage to stop discharging (normally 3.2 * 16)
3. Voltage to warn 3.15 * 16
4. Voltage to cut off 3.13 or 3.1 * 16
5. Voltage to stop charging: 3.45 * 16
6. All the rest just leave it for the inverter to manage (charging current).
When changing battery to lion with BMS connecting to the inverter (Luxpower) over CAN, inverter forget about voltage setting (it use voltage setting only if battery type is lead acid) and use SOC only. Therefore, I am thinking of using fake soc at the above voltages level . When battery cell reaching that voltage level, Soc will be 0-10% (below 3.2) or 100% (at 3.45).
I have tested this setting with esp32lolin and the inverter worked as expected. The only problem is fluctuation of SOC (as it based on voltage) during charge/discharge with high current.
That why I am thinking about a way to handle the in/out power overtime to bind the fakeSOC to.
Anyway, it is a trade off. I will go a bit more on this aspect to see how far it can go!
Best

**Reply by kommando828** [reply]
Maybe its simpler not to try and go for a complete replacement of the JK SOC but to set criteria where once these are triggered you intervene and make a correction and then give back the calc to the JK. I have not built my battery yet but plan to use a JK BMS, I know during oct to mar the SOC will be wrong as there is little to no solar but for the rest of the year the JK should work fine.

**Comment by Sleeper85** [comment]
> Yeah, you are all totally correct. However, as I have been observing the community in our places, most people just use JKBMS with battery and connect to the inverter as a lead acid battery (that use voltage as indicator for deciding charge/discharge). For the moment, this choice is mandated and prevail for all people here with DIY battery. JK-BMS is cheap and affordable, have some way to let the battery talk to the inverter is awesome but to invest some 100$ to buy a Smart Shunt (Victron/ Junctek ...) seems not the choice of many in our place. This is why I am still sleeping on the option of how to smoothly converting the voltage to some concept that inverter accept for treating the battery nicely with maximum durability with least investment (20-30$ or below on an esp32 and few other parts). Best Ngoc

Junctek shunts are affordable, you can find them for +/- $50. In addition, it will be simpler to use a shunt for multi-bms development.

Estimating the SoC based on voltage is possible but imprecise. It's easier to do this when the battery is at rest (no big discharging or charging). At the end of 2022 I programmed a script to control the JK-BMS SoC, this script was created based on the graph below and should only be used when the battery is at rest.

```python
if cell_v <= 2.9:
  print("SOC is very low")
elif cell_v <= 3.05:
  print("SOC : VOL : 0% - 2.9 ~ 3.05")
elif cell_v <= 3.16:
  print("SOC : 5% - 3.05 ~ 3.16")
elif cell_v <= 3.185:
  print("SOC : 10% - 3.16 ~ 3.185")
elif cell_v <= 3.22:
  print("SOC : 15% - 3.185 ~ 3.22")
elif cell_v <= 3.24:
  print("SOC : 20% - 3.22 ~ 3.24")
elif cell_v <= 3.26:
  print("SOC : 25% - 3.24 ~ 3.26")
elif cell_v <= 3.27:
  print("SOC : 30% - 3.26 ~ 3.27")
elif cell_v <= 3.275:
  print("SOC : 35% - 3.27 ~ 3.275")
elif cell_v <= 3.285:
  print("SOC : 40% - 3.275 ~ 3.285")
elif cell_v <= 3.29:
  print("SOC : 45% - 3.285 ~ 3.29")
elif cell_v <= 3.295:
  print("SOC : 50% - 3.29 ~ 3.295")
elif cell_v <= 3.3:
  print("SOC : 55% - 3.295 ~ 3.3")
elif cell_v <= 3.315:
  print("SOC : 60% - 3.3 ~ 3.315")
elif cell_v <= 3.32:
  print("SOC : 65% - 3.315 ~ 3.32")
elif cell_v <= 3.322:
  print("SOC : 70% - 3.32 ~ 3.322")
elif cell_v <= 3.325:
  print("SOC : 75% - 3.322 ~ 3.325")
elif cell_v <= 3.33:
  print("SOC : 80% - 3.325 ~ 3.33")
elif cell_v <= 3.34:
  print("SOC : 85% - 3.33 ~ 3.34")
elif cell_v <= 3.345:
  print("SOC : 90% - 3.34 ~ 3.345")
elif cell_v <= 3.365:
  print("SOC : 95% - 3.345 ~ 3.365")
elif cell_v <= 3.43:
  print("SOC : 100% - 3.365 ~ 3.43")
else:
  print("SOC is very high")
```

![LF280 LFP 280Ah battery discharge and charge curves](https://github.com/Sleeper85/esphome-jk-bms-can/assets/153944466/7029fbb9-3289-461a-97ba-c43dfb032410)

**Comment by Sleeper85** [comment]
> Maybe its simpler not to try and go for a complete replacement of the JK SOC but to set criteria where once these are triggered you intervene and make a correction and then give back the calc to the JK. I have not built my battery yet but plan to use a JK BMS, I know during oct to mar the SOC will be wrong as there is little to no solar but for the rest of the year the JK should work fine.

I completely agree with you, this is already what is being done now for SOC 100%.
100% is sent only when charging is complete.

**Reply by straga** [reply]
Also can use: https://www.ti.com/amplifier-circuit/current-sense/digital-power-monitors/overview.html with some shunt from ali.

**Reply by straga** [reply]
Some info: https://www.youtube.com/watch?v=maTqnSs_XAI

**Reply by paulsteigel** [reply]
Hi, my final solution is:
1. at low voltage below 3.2 in the range from 3.15 to 3.19, fixed a fake SOC at the level that inverter will stop charging (as it prefer SOC than voltage). There are two ways
+ Send can id 0x351 with discharge current set to 0 and this will recover to the BMS level when voltage is above 3.25 (in my case only happened when charging started in the next day), this will not trigger the alarm at inverter side
+ Leave the SOC at the level that Inverter will cut off discharging (in my case 20%) and recover to BMS SOC when voltage is above 3.25 as above
This trick will prevent the inverter to try discharging again when voltage recovers to 3.2 after the previous discharge stop.
2. the case of high voltage, the solution of @Sleeper85 is great as it can help reducing charging current so all are fine!

To @Sleeper85, I would recommend you to modify a bit the code on the voltage as some people may have 15 cells rather than 16 cells. Also, there is case that people do wiring other battery pack (different Ah) in parallel to the one with Can communication to charger or inverter. So adding a box to define number of battery pack and Ah can help converting current correctly in Can message to Charger/inverter
Best!
Ngoc

**Comment by paulsteigel** [comment]
> > Maybe its simpler not to try and go for a complete replacement of the JK SOC but to set criteria where once these are triggered you intervene and make a correction and then give back the calc to the JK. I have not built my battery yet but plan to use a JK BMS, I know during oct to mar the SOC will be wrong as there is little to no solar but for the rest of the year the JK should work fine.
> 
> I completely agree with you, this is already what is being done now for SOC 100%. 100% is sent only when charging is complete.

Hi, Sleeper
I have not come back to your repo and find a lot of updates have been in place.
I have forked your code and added with:
1. Under voltage protection to prevent the inverter to draw current till uvp trigger (3-3.2 V).
2. Over voltage protection to prevent the over charge.
3. discharging current limit by time range: this helpful for people who want higher discharge at certain time range. I dont know how to use the time entity of home assistant, so I make a simple input box.
4. Number of pack in parrallel
![image](https://github.com/Sleeper85/esphome-jk-bms-can/assets/8839597/0e9c6c09-a270-46f7-935c-cecb63f23781)

Today, I test your new config and found out that the current limit set to 0 (both discharge and charging) when soc is at 100% (a wrong jk soc), your 99% soc under can_id 0x355 become unusable to tell the inverter to continue charging until the UVP reached or nearly reached.
Regarding charging/discharging cycle: When charging this number is positive but at discharging, this number became weird. I don't really quite understand how inverter calculate the cycle but i doubt that it would expect a number relating to total charging cycle capacity and the battery capacity in Ah. I will test this to see how it would work.
![image](https://github.com/Sleeper85/esphome-jk-bms-can/assets/8839597/ff6b90f1-0965-4fbe-bb49-a4b81f1a2a4d)
I am using luxpower (SNA5000WM) and the canbus 0x70 is not working but 0x373 worked.
If possible, can you please be so kind cheching the auto voltage/current function on charg/discharge, it seems not quite working when firmware is flash at the middle of the charging process!
Many thanks

**Comment by kommando828** [comment]
I have now been running with a new JK-PB2A16S20P HW ver 15.1 SW ver 15.7 for 10 days and can make the following observations. I run in closed loop canbus with an SMA SI5048 with SI6048US firmware. They may be only applicable to this particular board and firmware so other users should check to see if what I tried works.

When I first turned the BMS on the % SOC the BMS reported was a good match between the V and the tables of LFP battery SOC vs resting V. I calibrated the V at rest and the A during charging as per instructions, they were different from the initial settings but not by much. The voltage reading on the app stayed in line with the SMA SI reported V, the A however got out of line the closer to Zero showing on the SI and was max a -1A at zero.

That -1.0A meant that the SOC decayed overnight when the draw was at its lowest compared to the on line LFP battery SOC tables. During the day when charging this meant that once the 55.2V charge target was met the JK would jump from 80% to 100%. So in 24 hrs it is losing 20% SOC. So the 40% SOC seen in the morning after seeing 100% previous day just prior to charging is probably 60%.

You could accept this if your charging is daily and is enough to trigger the 100% but 20% drift per day and 2 days of little to no sun got my SI to shut down at 15% reported SOC but the battery was still at 52V.

So then I went into the JK app and adjusted the battery capacity from 280AH to 281H during a quiet period when the battery was resting, the SOC % then jumped to a % matching the LFP battery SOC tables.

So assuming this also works with other JK BMS boards and firmware then you could have code to look for periods of battery resting and then change the battery AH capacity by 1AH.  The JK board would then reset the SOC to a more accurate %.

**Reply by paulsteigel** [reply]
Hi, 
Thanks for your sharing, this have been solved with reporting non-full SOC (i.e 98%) to the inverter when JK reports 100%. By doing this, we dont have to do anything with JK app.
Best,
Ngoc

**Comment by wolfgangr** [comment]
Just found this issue when I did a web search on "**jk bms kalman soc**".
Sad to see that "kalman" is not even mentioned at all at this page.

Just gleaned over some other documents yielded from said search phrase - and just have to smile on the impression how hard you diligent guys here appear to aim hard - just to miss the goal.

Given the stepped, over wide ranges nearly flat charge curve of LiFePO4 cells, I'm afraid there is no chance to get SOC by a simple backwards mapping from voltage.

As far as my crude scrutiny revealed

- the minimum model for LiFePO4 is an improved "Thevenin"-model that combines
-- "internal" voltage as function of SOC
-- ohmic resistance
-- fast RC-restistance modelling double layer capacitance
-- slow RC-resitance, modeling diffusion processes

As I understood (forgive if I'm wrong)
- precise models refine these parameters by temperature and aging and yield 0.3 % SOC
- more crude self learning Kalman stuff still is at 3% SOC
- everything else is > 20 % error if a pack is kept in middle SOC for a couple of cycles
- Kalman filtering is advertised for "low ressource use" and capable of fit on embedded systems.
Might it fit on ESP32?
- other AI, neuronal, massive data etc approaches are far beyond embedded systems


Or might it fit on a backend postprocessor getting the data by mqtt?

Anybody out there on the way to implement anything close to a serious SOC?

**Comment by Sleeper85** [comment]
@wolfgangr

If you are able to develop code for better JK BMS SOC calculation, do so and share it with the community 😉

**Reply by shvmm** [reply]
Couldn't agree more
